### PR TITLE
implementation of GetZoneByProviderID and GetZoneByNodeName for azure

### PR DESCRIPTION
**Pull request by yastij**
This is part of the #50926 effort

cc @luxas 

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by yastij** [comment]
whoops forgot an unused import

**Comment by FengyunPan** [comment]
Oh, Can you run unit test before creating this PR?

**Comment by yastij** [comment]
I did, I added it accidentally before my commit.

**Comment by yastij** [comment]
I removed the errors package import, ran tests, but I somehow managed to add it back before the commit :)

File: pkg/cloudprovider/providers/azure/azure_zones.go [hunk 1]
```diff
@@ -61,14 +60,31 @@ func (az *Cloud) GetZone() (cloudprovider.Zone, error) {
...
+
+	if err != nil {
+		return cloudprovider.Zone{}, err
+	}
+
+	if !exists {
+		return cloudprovider.Zone{}, nil
+	}
+	zone := cloudprovider.Zone{
+		FailureDomain: *(vm.AvailabilitySet.ID),
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_zones.go]: This is a full resource ID for the availability set, but fault domain should be a number.
Should be:
VirtualMachineProperties/VirtualMachineInstanceView/PlatformFaultDomain

Reference:
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/manage-availability
https://github.com/kubernetes/kubernetes/blob/74f4fda7beba6d6f9b56f92b08c3b11cd27397d4/pkg/cloudprovider/providers/azure/azure_zones.go#L53

Comment (yastij) [reply pkg/cloudprovider/providers/azure/azure_zones.go]: thanks for the hint.

**Comment by karataliu** [comment]
There's still another gap:
By default InstanceView property is null, you should fill in compute.InstanceView to 3rd parameter to get instanceView:
https://github.com/kubernetes/kubernetes/blob/5a25527b9b26521114fa5444966bc1a167704b7c/pkg/cloudprovider/providers/azure/azure_wrap.go#L49

Consider adding a direct call via az.VirtualMachinesClient.Get, since not everyone needs instanceView.

**Comment by yastij** [comment]
@karataliu - Adressed, thanks for the comments.

File: pkg/cloudprovider/providers/azure/azure_zones.go [hunk 2]
```diff
@@ -61,14 +62,36 @@ func (az *Cloud) GetZone() (cloudprovider.Zone, error) {
...
+	vm, err := az.VirtualMachinesClient.Get(az.ResourceGroup, string(nodeName), compute.InstanceView)
+
+	exists, err := checkResourceExistsFromError(err)
+	if err != nil {
+		return cloudprovider.Zone{}, err
+	}
+
+	if !exists {
+		return cloudprovider.Zone{}, nil
+	}
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_zones.go]: Node non-existence should be an error in this case. There's no need to call checkResourceExistsFromError for a check.

**Comment by yastij** [comment]
@karataliu - should be fine now, commits squashed.

File: pkg/cloudprovider/providers/azure/azure_zones.go [hunk 3]
```diff
@@ -61,14 +62,31 @@ func (az *Cloud) GetZone() (cloudprovider.Zone, error) {
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
 func (az *Cloud) GetZoneByProviderID(providerID string) (cloudprovider.Zone, error) {
-	return cloudprovider.Zone{}, errors.New("GetZoneByProviderID not implemented")
+	nodeName, err := splitProviderID(providerID)
```
Comment (andrewsykim) [inline pkg/cloudprovider/providers/azure/azure_zones.go]: Not a huge fan of this function name `splitProviderID`, but seems outside the scope of this PR :P

File: pkg/cloudprovider/providers/azure/azure_zones.go [hunk 4]
```diff
@@ -18,14 +18,15 @@ package azure
 
 import (
 	"encoding/json"
-	"errors"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"sync"
 
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
```
Comment (andrewsykim) [inline pkg/cloudprovider/providers/azure/azure_zones.go]: Can we group the `github.com/Azure` imports separate from the `k8s.io` imports and group `strconv` with the rest of the stdlib packges?

File: pkg/cloudprovider/providers/azure/azure_zones.go (hunk 3 above)
Comment (luxas) [reply pkg/cloudprovider/providers/azure/azure_zones.go]: yeah, not for now

File: pkg/cloudprovider/providers/azure/azure_zones.go (hunk 4 above)
Comment (luxas) [reply pkg/cloudprovider/providers/azure/azure_zones.go]: please fix

**Review by luxas — COMMENTED** [review]
@yastij can you please fixup this today?

File: pkg/cloudprovider/providers/azure/azure_zones.go (hunk 4 above)
Comment (yastij) [reply pkg/cloudprovider/providers/azure/azure_zones.go]: yup sure

**Comment by yastij** [comment]
@luxas @andrewsykim fixed :)

**Comment by andrewsykim** [comment]
@yastij any chance we can add tests for the `GetZoneBy*` methods? I'm not too familar with the azure code so I'm not too sure how testable it is.

**Comment by yastij** [comment]
@andrewsykim - the code is pretty trivial. I'll try to add a test after #51588 gets merged

**Comment by dims** [comment]
Do we need this for v1.8?

**Comment by luxas** [comment]
@dims yeah, I'd say so. I'd count this as a bug fix for the azure ccm

**Comment by dims** [comment]
@luxas - this seems to be waiting on https://github.com/kubernetes/kubernetes/pull/51588 which does not have the milestone v1.8 hence asking

**Comment by yastij** [comment]
@dims +1

**Comment by jdumars** [comment]
@brendandburns PTAL

**Comment by luxas** [comment]
(I interpret no objection there as a "ok-to-add-to-milestone" @jdumars)

**Comment by jdumars** [comment]
@luxas this PR is missing LGTM and approval, as well as 'approved-for-milestone' which has to come from the sponsor SIG
